### PR TITLE
Automated Long Term Memory Concept

### PR DESCRIPTION
**Pull request by mjnitz02**
Thought I would propose this idea that I had for an "Automated Long-Term Memory" solution. I use qvink in a style where I have it produce slightly more useful summaries over time, but this fills up the short-term context quicker. In longer chats this can lead to sort of a run-out where useful information tends to trickle off unless I'm actively thinking and tagging summaries. For me I usually forget to do this and would end up having to go back and try to determine `what is actually relevant`.

There's always a lot of "filler" messages no matter what anyone does, and the characters don't need to remember the filler, it's really not important overall in the long term (and the user will probably forget it anyways), but key moments really matter.

So I had an idea to work on this today and sort of got lost in the implementation and testing and thought I would at least throw it out as a concept for you to consider.

# What's in this PR
- Added a secondary default option that is better suited for prioritizing long term summarization flows
- Added the automated long term summarization system (see the README.md) for the main ideas but it has the following additions:
  - We collect all relevant message (either all, everything since the last long term message, or the last N messages)
  - Determine if we are using 1 or more chunks
  - Process chunks and get back indexes that should be flagged for long term
  - Flag the messages for long term storage
  - [Optional] We can unflag memories if a setting is selected
- We can also automatically run this every N messages for the user
- There is a small interesting tweak which alters the wording to try and determine what the LLM thinks is "important". After a lot of refinement this actually works pretty well, with High excluding a lot and Low being pretty generous about what is "important".
- There is a batch operation to allow the user to trigger the entire flow on everything
- On each message is a small Lightning Bolt Cloud icon. If you click this it assumes that message it the "most recent" message so it will trigger from there backwards in your chat. This can allow you to "reprocess" specific areas.
- Updated the `settings.html` to have the relevant settings and I tried to add really clear tooltips about what is happening here.

**Notably this doesn't really "change" Qvink_memory in much of a visual manner, it's more an automated backend functionality that is designed to streamline the concept of long term memory if a user is interested in doing that.**

# Testing
I tested this with GLM, DeepSeek, Claude and a few others. I worked to try and refine the prompts themselves (Using Clause Opus) so that they are appropriate for usage with Flash models (spent a while going back and forth with it trying to get it to think like a simpler model and tell me where it struggled). This led to prompts which are actually pretty reliable.

Overall I have actually been pretty please with the results, and its nice to not really have to "think" about the long term memory and just let the LLM determine what is actually important or not.

So this is mainly tuned towards "Flash" style models, but it works fine with the non-flash ones as well. It does generally select what I would consider the important message across several chats that I post processed and running it on new chats.

# Screenshots
Message Changes:
Screenshot 2026-03-18 at 4 28 59 PM

New Settings:
Screenshot 2026-03-18 at 4 28 42 PM

New Prompt for Auto Long-Term:
Screenshot 2026-03-18 at 4 28 49 PM

Summary Prompt Changes:
Screenshot 2026-03-18 at 4 33 10 PM
Screenshot 2026-03-18 at 4 33 15 PM

### CODE REVIEW

**Comment by mjnitz02** [comment]
@qvink So absolutely no strong encouragement to accept this PR at all. This was more a side project that I had been exploring in a local clone of the repo and I decided to toss it in a fork and open a PR for you to explore the idea. Out of all the memory extensions, I find this one actually works. the best for me overall, and this idea got rid of one of my personal pain points with the extension (now I can be lazier!).

**Comment by qvink** [comment]
Thanks so much for this! Before I can look at these changes, please submit to the `dev` branch instead as it has the most up-to-date features and changes, which you will need to account for. Most notably, the method used to send generations is completely different.

**Comment by mjnitz02** [comment]
@qvink Sure, give me a day or two and i can refactor it against that

File: index.js [hunk 1]
```diff
@@ -72,6 +72,10 @@ const remember_button_class = `${MODULE_NAME}_remember_button`
 const summarize_button_class = `${MODULE_NAME}_summarize_button`
 const edit_button_class = `${MODULE_NAME}_edit_button`
 const forget_button_class = `${MODULE_NAME}_forget_button`
+const automated_memory_auto_button_class = `${MODULE_NAME}_automated_memory_auto_button`
+
+// Count new messages since last automated LTM run
+var automated_memory_auto_message_counter = 0
```
Comment (mjnitz02) [inline index.js]: All of these top changes are basically just new variables needed to run things.

File: index.js [hunk 2]
```diff
@@ -3573,6 +3612,193 @@ async function remember_message_toggle(indexes=null, value=null) {
     }
     refresh_memory();
 }
+function collect_long_term_memory_summaries(scope, end_index) {
```
Comment (mjnitz02) [inline index.js]: This is the first of 3 new functions. It collects all the summarized messages and formats them in a way that the LLM can go through and identify the memories. This can then be re-associated to the initial messages

File: index.js [hunk 3]
```diff
@@ -3573,6 +3612,193 @@ async function remember_message_toggle(indexes=null, value=null) {
...
+        let memory = get_data(chat[i], 'memory');
+        if (!memory) continue;
+        let is_remembered = get_data(chat[i], 'remember') ? 'Yes' : 'No';
+        summary_lines.push(`Message ${i} | Long-term: ${is_remembered} | Summary: ${memory}`);
+        valid_indexes.push(i);
+    }
+
+    return { summary_lines, valid_indexes };
+}
+async function analyze_long_term_memory_summary_chunk(chunk_lines, chunk_indexes, prompt_template, importance_text, profile) {
```
Comment (mjnitz02) [inline index.js]: This function I just split out to make it a bit easier to follow, but it basically processes `1 chunk` of summaries to have the LLM get the importances.

File: index.js [hunk 4]
```diff
@@ -3573,6 +3612,193 @@ async function remember_message_toggle(indexes=null, value=null) {
...
+
+    if (content.trim().toUpperCase() === 'NONE') {
+        return [];
+    }
+
+    let parsed_numbers = content.match(/\d+/g)?.map(Number) || [];
+    let chunk_valid_set = new Set(chunk_indexes);
+    return parsed_numbers.filter(n => chunk_valid_set.has(n));
+}
+async function automated_long_term_memory(scope_override=null, end_index=null) {
```
Comment (mjnitz02) [inline index.js]: This is the main `automated_long_term_memory` concept. Basically we can analyze messages in one of a few ways. Depending on the flow we analyze 1 or more "chunks" of messages at a time to get the overall context. In the `settings.xml` I tried to illustrate how to use each option.

File: index.js [hunk 5]
```diff
@@ -3573,6 +3612,193 @@ async function remember_message_toggle(indexes=null, value=null) {
...
+        return;
+    }
+
+    // Step 2: Determine chunking — when processing all with a non-'all' scope, split into chunks of last_n.
+    // 'all' scope sends everything in a single batch; 'new' and 'last_n' chunk by last_n size.
+    let chunk_size = get_settings('automated_memory_last_n') || 50;
+    let is_process_all = scope_override === 'all';
+    let scope_allows_chunking = get_settings('automated_memory_scope') !== 'all';
+    let exceeds_chunk_size = summary_lines.length > chunk_size;
+    let use_chunking = is_process_all && scope_allows_chunking && exceeds_chunk_size;
```
Comment (mjnitz02) [inline index.js]: Broke all these booleans up just to make the conditions clear. Technically they could be a giant 1-liner, but thats sort of unreadable.

File: index.js [hunk 6]
```diff
@@ -3573,6 +3612,193 @@ async function remember_message_toggle(indexes=null, value=null) {
...
+
+    for (let c = 0; c < chunks.length; c++) {
+        let chunk = chunks[c];
+
+        if (use_chunking) {
+            toastr.clear($progress_toast);
+            $progress_toast = toastr.info(
+                `Processing batch ${c + 1} of ${total_chunks} (${chunk.lines.length} messages)...`,
+                MODULE_NAME_FANCY,
+                { timeOut: 0, extendedTimeOut: 0, tapToDismiss: false }
```
Comment (mjnitz02) [inline index.js]: Throughout the summarization process (because automated LTM only run sparingly) the user gets toast notifications that tell them what is going on (blue/green: running/finished, yellow: warnings, red: failures).

As long as the backend is working ok people should usually only see blue/green. If your provider flakes you might see the other colors.

File: index.js [hunk 7]
```diff
@@ -4042,6 +4268,21 @@ async function auto_summarize_chat(skip_initial_delay=true) {
...
+    automated_memory_auto_message_counter++;
+    let interval = Math.max(10, get_settings('automated_memory_auto_interval') || 25);
+    debug(`Auto LTM counter: ${automated_memory_auto_message_counter} / ${interval}`);
+
+    if (automated_memory_auto_message_counter >= interval) {
+        automated_memory_auto_message_counter = 0;
+        log('Auto LTM interval reached, running automated long-term memory analysis...');
+        await automated_long_term_memory();
+    }
+}
```
Comment (mjnitz02) [inline index.js]: I added this after the original implementation, and basically it sort of mirrors the existing auto triggers, but it counts messages to determine if the Automated Long Term Memory should be automatically run (only if enabled in settings.html).

File: settings.html [hunk 8]
```diff
@@ -310,6 +310,55 @@ <h4 class="textAlignCenter">Long-Term Memory Injection <i class="fa-solid fa-inf
                     </select>
                 </label>
 
+                <!-- Automated Long-term memory -->
```
Comment (mjnitz02) [inline settings.html]: This whole section just adds the new settings to the bottom of the settings to allow the user to control how this is run. I tried a few different layouts, and tried to sort of mirror the existing layouts but keep it intuitive. The tooltips are pretty verbose, but I wanted to try and make it extremely clear exactly what was happening when you had things selected so there wasn't any room for confusion.

**Review by mjnitz02 — COMMENTED** [review]
@qvink So I rebased against `dev` and removed the extra prompt toggling stuff for now. This is only the `Automated Long Term Memory` concept.

The `tl;dr` being you are deferring all or chunks of your memory summaries to an LLM to have it determine if something is "critical enough" to be flagged as a long term memory.

I haven't had a chance to install the dev branch in my environment yet and test it. I'll try to do that maybe tomorrow if I get some time. Most all of the code was new, but I squashed the commits to make rebasing easier. I also updated the connection profiles to use the new flow, so I'll try to give that a test.

File: settings.html (hunk 8 above)
Comment (qvink) [reply settings.html]: I am a fan of long tooltips :)

**Comment by qvink** [comment]
Thanks again for the effort, I like the direction this is going! It's not quite in a testable state yet so I can't comment on function, but I can share my thoughts so far based on what I see here.

1. Given that this task is quite different than summarizing, I wouldn't be surprised if users will want to use a separate connection profile for it. I know I'd want to use a model more suited for consistent, structured output. 
2. This appears to be intended to work alongside manually marked long-term memories. If that is the case, we should differentiate between manually-marked and auto-marked summaries, especially for the `Edit Memory` interface. It's definitely something I'd want to filter/select for when making broad edits.
3. The "Process all Memories" button should instead be a bulk action in the `Edit Memory` interface so the user can make use of the filtering and manual selection of messages.
4. The chunking mechanism needs to be able to adjust to the max allowed context of the selected connection profile as that is not something the user should have to estimate.
5. I may be wrong but It doesn't look like this handles what would happen if both a summary and an auto-analysis were to happen at the same time. Or multiple auto-analyses for that matter. I'm thinking that these generations should probably be incorporated into the `SummaryQueue` so that it respects the user's `parallel summaries` and `time delay`settings. That way we can also use the progress bar instead of toasts.
6. I'm against hard-coded prompts wherever possible (aside from defaults) as prompt efficacy is extremely model-dependent. I know the `{{importance}}` macro is optional, but it's very much baked into the config here. For the sake of simplicity, I think it's best to just use one of those importance statements in the default prompt and leave out the macro.
7. Along the same lines, the format of summaries sent to the model is completely hard-coded. 

Regarding the last points: At some point I'll be generalizing the summary prompt editor so it can be used elsewhere, and at that time I'll be able to use it for this prompt as well. Using that we could allow customizing the format of the summaries and adding your importance options. You just might need to wait a bit for that.

**Comment by mjnitz02** [comment]
@qvink These are all great suggestions. Give me a few days to take a look at incorporating them. I hadn't really thought of the idea of integrating it mainly into the messages viewer, but that makes a lot of sense, along with all the profile and prompt suggestions. I'll incorporate this feedback into it.

**Comment by mjnitz02** [comment]
@qvink Sorry for the long delays. Haven't had a ton of time for stuff. I made some changes but still a few more to go. Also I fixed the connection_profile bug that was making it not runnable on the latest dev branch.

## Areas that have changed.
> This appears to be intended to work alongside manually marked long-term memories. If that is the case, we should differentiate between manually-marked and auto-marked summaries, especially for the Edit Memory interface. It's definitely something I'd want to filter/select for when making broad edits.

I added an `auto_remember` setting and filters and integrated this. This actually simplified a lot and the "Process all memories" button and strange settings was able to be removed as it was redundant.

> The "Process all Memories" button should instead be a bulk action in the Edit Memory interface so the user can make use of the filtering and manual selection of messages.

Updated this to use this location for handling user-directed auto tagging

> The chunking mechanism needs to be able to adjust to the max allowed context of the selected connection profile as that is not something the user should have to estimate.

I thought about this a lot after integrating it into the Memory Editor, and realized that chunking really isn't needed. The user can either manually select chunks, or use like `last_n` messages and sort of scroll through if they want hyper-specific chunks. If they select everything, they probably want it all analyzed as one big block looking for the absolutely most critical things. So chunking felt sort of wrong after enabling everything in the Memory Editor and I pulled it out entirely.

> I may be wrong but It doesn't look like this handles what would happen if both a summary and an auto-analysis were to happen at the same time. Or multiple auto-analyses for that matter. I'm thinking that these generations should probably be incorporated into the SummaryQueue so that it respects the user's parallel summaries and time delaysettings. That way we can also use the progress bar instead of toasts.

I changed the connection over to using the SummaryQueue, but I'm not entirely sure if it addresses this or not. I can try to do some testing here. I'll look into this more.

> I'm against hard-coded prompts wherever possible (aside from defaults) as prompt efficacy is extremely model-dependent. I know the {{importance}} macro is optional, but it's very much baked into the config here. For the sake of simplicity, I think it's best to just use one of those importance statements in the default prompt and leave out the macro.

Yea, this made a lot of sense. I just used the `medium` text as the default and got rid of all the `{{importance}}` related logic and code.

> Along the same lines, the format of summaries sent to the model is completely hard-coded.

This one I wasn't fully sure about. We need the model to return a list of numbers. Would making this a secondary exposed prompt be better? Each summary sort of needs to be iterated through and attached to a number and the LLM needs to be instructed to return those numbers. I wasn't sure how to allow the user to define this loop.

## Things I still need to look into
- Given that this task is quite different than summarizing, I wouldn't be surprised if users will want to use a separate connection profile for it. I know I'd want to use a model more suited for consistent, structured output.
  - Is it worth just creating a fully separate and secondary `connection_profile` then? It took me a bit to fully understand this, but I think I get the gist of how it works now.
- That way we can also use the progress bar instead of toasts.

File: index.js [hunk 9]
```diff
@@ -88,6 +92,16 @@ Following is a history of messages for context:
 Following is the message to summarize:
 {{message}}
 `
+const default_automated_memory_prompt = `You are a narrative analysis assistant. Below is a list of message summaries from a fictional roleplay. Each line has this format: Message NUMBER | Long-term: Yes/No | Summary: TEXT
+
+{{summaries}}
```
Comment (mjnitz02) [inline index.js]: @qvink This was the one area you mentioned I wasn't sure how best to handle. Summaries are defined in short as:
```
summary_lines.push(`Message ${i} | Long-term: ${is_remembered} | Summary: ${memory}`);
```

So that as they iterate through each message it gives the LLM the index and the associated summary for it to review. This specific line is sort of hardcoded, and it mentions the format in the prompt. So the user can change a lot, but I wasn't really sure how to allow them to change this sort of thing since the iteration itself is done in the backend.

Comment (qvink) [reply index.js]: > This specific line is sort of hardcoded, and it mentions the format in the prompt

That's just it - the prompt mentions the specific format, and you can change the prompt, but not the format.

My initial thought is providing a customizable template for the user to edit, using macros to insert each dynamic portion. This would then be repeated for every summary. For example, the user could do something like:
`Message: {{index}} | Long-term: {{is_remembered}} | Summary: {{summary}} \n`
or if they don't want to include the long-term marker or a different separator, just:
`Index: {{index}}, Content: {{summary}} \n`

I just implemented something very similar to this over in https://github.com/qvink/SillyTavern-MessageSummarize/pull/37, where I am making it possible to customize the format of summaries injected into context.

**Comment by qvink** [comment]
> Sorry for the long delays

No rush at all!


> I added an auto_remember setting and filters and integrated this

Awesome! I like the color choice for auto-marked summaries, looks like it works. I might want to start condensing the `Edit Memory` interface filters, but that's a me problem for later.


> I thought about this a lot after integrating it into the Memory Editor, and realized that chunking really isn't needed

The issue of context size remains. If the user wants more messages analyzed than fits their context window, then you simply can't without chunking. I don't think this is a necessary feature, but if you don't want to limit the number of messages analyzed you'd have to chunk them according to the connection profile's max context size.

> I changed the connection over to using the SummaryQueue, but I'm not entirely sure if it addresses this or not

It should, because the SummaryQueue automatically respects the user's parallel request and delay settings for anything that goes through it. Though I also haven't tested any edge cases.

> Is it worth just creating a fully separate and secondary connection_profile then? It took me a bit to fully understand this, but I think I get the gist of how it works now.

I'm also on the fence. On one hand, it's yet another config option. On the other hand, I can totally see someone wanting to use a different model or even just a different completion preset (e.g. to adjust the max response length). Not necessary for sure, so I'd say it's your call for now.

> That way we can also use the progress bar instead of toasts.

hmm ya I might need to make the summary queue progress bar have variable text passed in when calling it. Maybe you don't worry about that, I can add it in later.

**Comment by belfon** [comment]
1) This works great, almost "flawlessly" (and by that I mean meets my expectations and I can just let it run without intervention) via Gemma4 31B locally (Q5).

2) I did change the prompt slightly to better serve my needs and work better. Feel free to see my changes and incorporate or not, but I think this selects better. I added a "locations" line, "smut" line (NSFW I guess, but I worded classy), and a "compare identical" line for increased functionality. Also added a thing about bullet points but that's specific for my own formatting I use.

```
You are a narrative analysis assistant. Below is a list of bullet point message summaries in chronological order in media res from a fictional roleplay. Each line has this format: Message NUMBER | Long-term: Yes/No | Summary: TEXT

{{summaries}}

Your task: decide which of these messages should be preserved as long-term memories. A message is worth preserving if it introduces important characters, advances the plot, contains a key decision or revelation, contains a location or setting change from immediate previous summaries, or establishes facts the story depends on. Do NOT include messages that are routine actions, filler, minor reactions, or repeat information already covered by another remembered message. Sex/smut memories can be worth preserving if it contains info about initiation, consent, climax, or relationship development. If multiple messages in the group seem to contain identical or near-identical information, you compare and decide which is superior (contains higher quality and/or more information as defined earlier) and select to preserve only the most superior while not including the inferior.

{{importance}}

Respond with ONLY the message numbers as a comma-separated list. Do not include any other text, explanation, or formatting. If no messages qualify, respond with the single word NONE.
```
